Add state option to git_config module

ari-stark opened the pull request:
State present/absent option works like set/unset option for 'git config'.

##### SUMMARY
Adding a state option to git_config module. State can be 'present' or 'absent'.
Works like 'git config' when 'present' and like 'git config --unset' when 'absent'.

Fixes #28567

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
git_config

ari-stark committed:
Add state option to git_config module

State present/absent option works like --set/--unset option for 'git config'.

mgedmin commented on lib/ansible/modules/source_control/git_config.py:
```diff
@@ -212,13 +230,19 @@ def main():
 
+    if unset:
+        args.insert(len(args) - 1, "--" + unset)
+        new_value = ''  # No new value when unset
```
This will still result in a command like `git config --unset foo ''`.  According to the git config man page, that extra argument is a "value_regex" and its presence means only those values matching this regex will be unset.

Luckily the empty string is a regex that matches everything, so it all works out fine in the end.

ari-stark replied:
You're right. I didn't see it.
Taken into account.

mgedmin reviewed (approved):
LGTM, but I get the impression Ansible maintainers these days want PRs to be accompanied by tests .

ari-stark committed:
Change git_config to avoid useless parameter passed to git command

When unsetting value, command was : git config --unset foo ''.
Now command is : git config --unset foo.

ari-stark commented:
I'll try to make some tests, but I'm new to Python and still have a lot to learn.

mgedmin commented:
There are also [integration tests](https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html#testing-integration) that are written as Ansible roles and can run in Docker containers.

Since for me it's always easier to understand from an example, I'd suggest taking a look at [`test/integration/targets/authorized_key/`](https://github.com/ansible/ansible/tree/devel/test/integration/targets/authorized_key).  You can run it with 

    test/runner/ansible-test integration -v authorized_key --docker 

Since I'm myself a drive-by contributor rather than a core Ansible developer, I cannot say if core devs would prefer unit or integration tests, or if they'd be happy to merge this as-is without any tests.

ari-stark commented:
Thanks a lot for this links. I just took a look at it, but it seems promising.
Whatever devs' preferences, I'll probably try both ways to test: I need to learn them both.
I'll begin with integration tests as it seems easier to me.

ari-stark committed:
Add some integration tests to git_config module

ari-stark committed:
Add missing aliases file

ari-stark committed:
Change set up method

Using git command seems to cause troubles on some OS : changing config
by changing '.gitconfig' file.

ari-stark committed:
Remove some distros from tests pool

Git is not installed or is out of date on these distros.

ari-stark committed:
Fix aliases to skip tests on centos6

ari-stark committed:
Refactor tests of the git_config module

ari-stark committed:
Add use case when state=absent and value is defined

ari-stark commented:
I added integration tests. I think it's a full coverage for the `state` option.
Is that OK ?

mgedmin reviewed (approved):
LGTM

resmo reviewed (approved):
shipit

resmo commented:
rebuild_merge

ju2wheels commented:
The integration tests for this seem to be leaving a `git` config behind which is affecting other integration tests. Im seeing the HTTP proxy foo that was added above in my integration tests and its affecting `git` clones done after the git_config integration tests im assuming since its not cleaning up any added global `git` config, see output of latest test run for #50829 at https://app.shippable.com/github/ansible/ansible/runs/111778/summary/console .

cc @maxamillion @resmo @mgedmin
